iceberg: add support for `write.{data,metadata}.path` table properties

wdberkeley opened the pull request:
This adds support for the `write.metadata.path` and `write.data.path` table properties, which for some catalogs specify where metadata and data, respectively, should be written. In their absence, the default is `{table location}/metadata` and `{table location}/data`, respectively.

I have tested this manually with the databricks_test.py Ducktape test. I'm planning to follow up with some unit tests for `write.data.path`. This change includes unit tests for `write.metadata.path`.

Fixes CORE-12152

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes

### Improvements

* Adds support for the Iceberg table properties `write.metadata.path` and `write.data.path`. When an Iceberg catalog defines these properties, Redpanda will use them to determine where to write Iceberg table metadata and data, respectively, instead of using default locations based on the table location.

wdberkeley committed:
datalake: copy table properties into table_info

Some catalogs require respecting properties returned by load_table. This
is the first step towards that.

wdberkeley committed:
datalake: respect `write.data.path`

Some catalogs require writing data to the location specified by the
`write.data.path` property. This changes the data location to respect
this property when it is present.

andrwng commented on src/v/datalake/catalog_schema_manager.h:
```diff
@@ -40,6 +40,7 @@ class schema_manager {
         iceberg::schema schema;
         iceberg::partition_spec partition_spec;
         iceberg::uri location;
+        std::optional<iceberg::table_properties_t> properties;
```
Another thought would be to exclusively pull out just a `std::optional<iceberg::uri> write_data_path` rather than copying so many values (seems like even the metadata path isn't derived from `table_info`? though i might just be missing it)

wdberkeley replied:
My instinct is that we'll need to support more properties over time so I grabbed the whole thing.

The metadata path is derived from the table info, too.

andrwng reviewed (dismissed):
Functionally lgtm, though the amount of metadata passed around in the catalog schema manager can maybe be trimmed down?

nvartolomei commented on src/v/datalake/catalog_schema_manager.cc:
```diff
@@ -20,6 +20,21 @@
 namespace {
+
+std::optional<iceberg::table_properties_t>
+copy_properties(const std::optional<iceberg::table_properties_t>& props) {
```
This method is a bit of a footgun. Today, because of the choice of the value in the map it does indeed to a copy. But imagine tomorrow someone goes and modifies table_properties_t to become a `std::map<ss::string, ss::shared_ptr<ss::string>>` (for the sake of example). Now, this method doesn't do copying anymore. And the compiler will remain silent.

I see 2 nicer approaches (there might be others ofc but i'll go with what comes now to mind):
1) You don't use the type alias and instead hard code the map. That way, this method will break and whoever will be doing the refactor will have to understand if they can just change the typing of the method or they have to refactor the method too to keep the old behavior (deep copying).
2) Make table properties sort of a class and put the method on it. That way, whoever is changing the class will be more likely to consider whether class methods retain their behavior or not. Compared to current state of the things where there isn't much relationship between the alias and this method.

wdberkeley replied:
Good point. I'll do option 1.

nvartolomei commented on src/v/datalake/catalog_schema_manager.cc:
```diff
@@ -20,6 +20,21 @@
 namespace {
+
+std::optional<iceberg::table_properties_t>
+copy_properties(const std::optional<iceberg::table_properties_t>& props) {
```
Optional as parameters, especially as only parameters, are a bit of a smell. Consider getting rid of it and instead call it with https://en.cppreference.com/w/cpp/utility/optional/transform.

I.e. `.properties = it->second.properties.transform(copy_properties),`

wdberkeley replied:
True. I did it that way to make it ergonomic at the callsite, but transform is more ergonomic. Good to see some functional-looking C++ code 🙂 .

nvartolomei reviewed (dismissed):
This is surprisingly less cumbersome than I expected. Thanks for not overcomplicating it!

@wdberkeley please add a test for this. I figured out the following works with vanilla spark iceberg plugin so I reckon it will work with rest catalog too.

1. Copy this test https://github.com/redpanda-data/redpanda/blob/a1eb2daa3908d72915822ec6b56c4527b14aec1f/tests/rptest/tests/datalake/datalake_e2e_test.py#L804
2. Produce a record to the topic then wait for table to be created.
3. Run the following spark query: 
```
f"""alter table {table_name} set TBLPROPERTIES ('write.data.path' = "{table.location()/foobarbaz}")"""
4. Write some additional data
5. Check that all `files = spark.run_query_fetch_all(f"select file_path from {table_name}.files")` but one (or whatever you figure out is good enough) respect the new location.
```

Similar for metadata.

This is just a sketch. Might not work but please try 🙇‍♂️

nvartolomei commented on src/v/iceberg/merge_append_action.cc:
```diff
@@ -27,18 +27,38 @@
+      = "write.metadata.path";
+
+    if (table.properties.has_value()) {
+        auto it = table.properties->find(ss::sstring(write_metadata_path_prop));
```
Was curious if this allocation can be avoided and it likely can if you use custom hasher for the container, like here 

see this https://github.com/redpanda-data/redpanda/blob/157249bcd3eddeea7f9b154722ede92fe9a6bdbf/src/v/datalake/tests/record_generator.h#L64

and this https://github.com/redpanda-data/redpanda/blob/e0a6210d4ac01eaddfd67bbbe198b5705433ec9f/src/v/datalake/tests/record_generator.cc#L98

---

can you give it a try?

wdberkeley replied:
Works. Neat.

nvartolomei commented:
@wdberkeley please mark this for 25.1 backport and add an improvement release note. This introduce a slight change in behavior now and though I don't think it will cause problems to anyone it would be good to have it there in case it does actually cause some problems.

@andrwng I'm for merging and backporting this before the custom test I asked for is written. It passes databricks test so it likely does the right thing when these properties are set. Let's do a manual validation (check the logs where data is uploaded) and ship it.

wdberkeley committed:
iceberg: respect write.metadata.path

Some catalogs require clients to respect the `write.metadata.path` table
property. This changes the metadata location to respect this property
when it is present.

dotnwat commented on src/v/datalake/catalog_schema_manager.cc:
```diff
@@ -20,6 +20,21 @@
+// NB: Use the concrete type not the alias iceberg::table_properties_t
+// so changes to the alias can't silently destroy the deep copy semantics
+// of this function.
+chunked_hash_map<ss::sstring, ss::sstring, sstring_hash, sstring_eq>
+copy_properties(
+  const chunked_hash_map<ss::sstring, ss::sstring, sstring_hash, sstring_eq>&
+    props) {
+    chunked_hash_map<ss::sstring, ss::sstring, sstring_hash, sstring_eq> result;
+    for (const auto& [key, value] : props) {
+        result.emplace(key, value);
+    }
+    return result;
+}
```
i must be missing something here. isn't chunked_hash_map copyable?

wdberkeley replied:
Nope. I think the underlying storage type `fragmented_vector` has a `copy` method but is not copyable.

dotnwat replied:
i think it may have made more sense to add a `copy()` method to `chunked_hash_map`?

nvartolomei replied:
@dotnwat I'm curious about your thoughts on my comment here https://github.com/redpanda-data/redpanda/pull/26426#discussion_r2140880456. Conceptually here we're interested in deep copying. A generic copy method on chuncked_hash_map as well as a copy constructor wouldn't have these properties. Feels like having a copy_properties method is still useful even if the underlying container provides a shallow copying mechanism.

Wonder if you have a better take on the problem I outlined in the comment above, in this concrete case as well as in general case.

dotnwat replied:
afaict the reason we can't copy chunked_hash_map is because the underlying anker segmented map tries to invoke copy constructor on our `fragmented_vector` type (the m_values member). one option would be to change anker copy constructor to build the copy from an iterator rather than using copy constructor.

the reason we don't allow copying of fragmented_vector is that we don't want to accidentically introduce the costs of large copies (same with iobuf). it's entirely possible that we haven't made the change is suggested because we want that same property for chunked_hash_map.

but, both iobuf and fragmented_vector have a copy methods that afaict are semantically equivalent to copy constructor. so why not provide chunked_hash_map with a similar interface?

as far as deep vs shallow copy goes, i get your point, but that is entirely dependent on the use case of the container. in this particular case a map from concrete string to concrete string is being copied, so i don't see any concern that might arise if the container were storing pointers.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--aa51c9b9-d9d8-46e7-a4e7-aee524815577-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67237">build#67237</a></summary>


| test_class               | test_method          | test_arguments                                                                                                                      | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                      |
|:-------------------------|:---------------------|:------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------|
| RandomNodeOperationsTest | test_node_operations | {"cloud_storage_type": 2, "enable_failures": true, "mixed_versions": false, "with_chunked_compaction": true, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67237#01976135-845f-48ae-b051-ed4a47396583 | FLAKY         | 15/21    | upstream reliability is '100.0'. current run reliability is '71.42857142857143'. drift is 28.57143 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

vbotbuildovich commented:
/backport v25.1.x
